File: common/src/main/resources/assets/mcui/themes/hex2/scripts/menu_support.lua
--- @module menu_support
local menu = {}

local theme = require 'theme'
local util = require 'util'
local wl = require 'widgetlib'

local menu_frag = theme.readWidget("mcui.hex2:menu_parent")
local page_frag = theme.readWidget("mcui.hex2:menu_page")

--- Adds a paged menu to given widget
--- @param parent Widget|string
--- @return Widget
function menu.addMenu(parent)
    local r = theme.loadWidget(parent, menu_frag)
    return r or error('Unable to load menu')
end

--- Adds a page to the menu
--- @param menuRoot Widget|string
--- @param key string
--- @param parentKey? string|nil
--- @return Widget
function menu.addMenuPage(menuRoot, key, parentKey)
    local r = theme.loadWidget(menuRoot, page_frag, {
        pageKey = wl.tstatic(key),
        parentPageKey = wl.tstatic(parentKey or ''),
    })
    return r or error('Unable to load page ' .. key)
end

--- Adds a button to navigate to the parent page
--- @param page Widget
--- @param args? buttonArgs
function menu.addBackButton(page, args)
    wl.loadButton(
        page.getChildByName('content') or page, -- in case we want to give the content itself
        util.merge(
            {
                yPos = -10,
                width = 160,
                label = wl.tstatic('format("mcui.screen.back")', 'STRING', true),
                onClick = function (self)
                    self.setVariable('isHovered', wl.tstatic(false))
                    return page.extra.close()
                end
            }, args or {}
        )
    )
end

--- Adds a button to navigate to another page
--- @param menuRoot Widget
--- @param onPageContent Widget
--- @param toPage string
--- @param label? fun(id: string): string
--- @param args? buttonArgs
function menu.addButtonToPage(menuRoot, onPageContent, toPage, label, args)
    wl.loadButton(
        onPageContent,
        util.merge(
            {
                key = toPage,
                label = wl.tstatic((label and label(toPage)) or '"' .. label .. '"', "STRING", true),
                xPos = -80,
                width = 160,
                onClick = function ()
                    return menuRoot.extra.open(toPage)
                end,
            }, args or {}
        )
    )
end

return menu

File: hex_ts/src/main/ts/menu_support.ts
import type {Widget} from "../../typings/Widget";
import {type ButtonArgs, loadButton, typedStaticValue} from "./widget_lib";

const menu_frag = theme.readWidget("mcui.hex-ts:menu_parent");
const page_frag = theme.readWidget("mcui.hex-ts:menu_page");

export interface Menu extends Widget {
    extra: {
        open: (this: void, pageKey: string) => boolean;
    }
}

/**
 * Adds a paged menu to given widget
 */
export function addMenu(parent: Widget | string): Menu {
    const r = theme.loadWidget(parent, menu_frag) as Menu || error(`Unable to load menu`);

    r.extra.open = (pageKey) => {
        const trimmedPageKey = pageKey.replaceAll('"', '');
        r.allChildren.forEach(childElement => {
            if (childElement.name === 'menu_page') {
                const child = childElement as Widget;
                const childKey = child.getVariable('pageKey').expression.replaceAll('"', '');
                child.setVariable('isOpen', typedStaticValue(childKey === trimmedPageKey));
            }
        })

        return true;
    }

    return r
}

export interface MenuPage extends Widget {
    extra: {
        close: (this: void) => boolean;
    }
}

/**
 * Adds a page to the menu
 */
export function addMenuPage(menuRoot: Widget | string, key: string, parentKey?: string): MenuPage {
    const r = theme.loadWidget(menuRoot, page_frag, {
        pageKey: typedStaticValue(key, "STRING"),
        parentPageKey: typedStaticValue(parentKey || '', "STRING"),
    }) as MenuPage || error(`Unable to load page ${key}`);
    r.extra.close = () => {
        (r.parentElement as Widget).extra.open?.(r.getVariable('parentPageKey').expression)
        return true;
    }
    return r;
}

/**
 * Adds a button to navigate to the parent page
 */
export function addBackButton(page: MenuPage, args?: ButtonArgs): void {
    const content = page.getChildByName('content') as Widget || page; // in case we want to give the content itself
    loadButton(
        content,
        {
            yPos: -10,
            width: 160,
            label: typedStaticValue('format("mcui.screen.back")', 'STRING', true),
            onClick: function (this: Widget) {
                this.setVariable('isHovered', typedStaticValue(false, "BOOLEAN"));
                return page.extra.close();
            },
            ...(args ?? {})
        }
    );
}

/**
 * Adds a button to navigate to another page
 */
export function addButtonToPage(
    menuRoot: Menu,
    onPageContent: Widget,
    toPage: string,
    label?: ((id: string) => string) | null,
    args?: Partial<ButtonArgs>
): void {
    loadButton(
        onPageContent,
        {
            key: toPage,
            label: typedStaticValue(
                (label ? label(toPage) : `"${toPage}"`),
                "STRING",
                true
            ),
            xPos: -80,
            width: 160,
            onClick: function () {
                return menuRoot.extra.open(toPage);
            },
            ...(args ?? {})
        }
    );
}
